File: 0229-majority-element-ii/0229-majority-element-ii.py
class Solution:
    def majorityElement(self, nums: List[int]) -> List[int]:
        d = {}
        n = len(nums)
        ans = []
        for i in nums:
            if i in d:
                d[i] += 1
                if d[i] > n//3 and i not in ans:
                    ans.append(i)
            else:
                d[i] = 1
                if d[i] > n//3:
                    ans.append(i)
        return ans
                

File: 0229-majority-element-ii/0229-majority-element-ii.java
class Solution {
    public List<Integer> majorityElement(int[] nums) {
        HashMap<Integer, Object[]> map = new HashMap<>();
        List<Integer> ans = new ArrayList<>();
        for(int num : nums){
            if(map.containsKey(num)){
                int frequency = (int)map.get(num)[0];
                boolean present = (boolean)map.get(num)[1];
                frequency++;
                if(frequency > nums.length/3 && present == false){
                    ans.add(num);
                    map.put(num, new Object[] {frequency, true});
                }
                else{
                    map.put(num, new Object[] {frequency, present});
                                    
                }
            }
            else{
                int frequency = 1;
                if(frequency > nums.length/3){
                    ans.add(num);
                    map.put(num, new Object[] {frequency, true});
                }
                else{
                    map.put(num, new Object[] {frequency, false});
                }
            }
            
        }
        return ans;
        
    }
}